src/dp/distributions.rs

```diff
@@ -0,0 +1,603 @@
+    let summand = sigma.pow(2) / t.clone();
+    // compute probability of accepting the laplace sample y
+    let prob = |term: Ratio<BigUint>| term.pow(2) * (sigma.pow(2) * BigUint::from(2u8)).recip();
+
+    loop {
+        let y = sample_discrete_laplace(&t, rng);
+
+        // absolute value without type conversion
+        let y_abs: Ratio<BigUint> = BigUint::new(y.to_u32_digits().1).into();
+
+        // unsigned subtraction-followed-by-square
+        let prob: Ratio<BigUint> = if y_abs < summand {
+            prob(summand.clone() - y_abs)
+        } else {
+            prob(y_abs - summand.clone())
+        };
+
+        if sample_bernoulli_exp(&prob, rng) {
+            return y;
+        }
+    }
+}
+
+/// Samples `BigInt` numbers according to the discrete Gaussian distribution with mean zero.
+/// The distribution is defined over the integers, represented by arbitrary-precision integers.
+/// The sampling procedure follows [[CKS20]].
+///
+/// [CKS20]: https://arxiv.org/pdf/2004.00010.pdf
+#[derive(Clone, Debug)]
+pub struct DiscreteGaussian {
+    /// The standard deviation of the distribution.
+    std: Ratio<BigUint>,
+}
+
+impl DiscreteGaussian {
+    /// Create a new sampler from the Discrete Gaussian Distribution with the given
+    /// standard deviation and mean zero. Errors if the input has denominator zero.
+    pub fn new(std: Ratio<BigUint>) -> Result<DiscreteGaussian, DpError> {
+        if std.denom().is_zero() {
+            return Err(DpError::ZeroDenominator);
+        }
+        Ok(DiscreteGaussian { std })
+    }
+}
+
+impl Distribution<BigInt> for DiscreteGaussian {
+    fn sample<R>(&self, rng: &mut R) -> BigInt
+    where
+        R: Rng + ?Sized,
+    {
+        sample_discrete_gaussian(&self.std, rng)
+    }
+}
+
+impl DifferentialPrivacyDistribution for DiscreteGaussian {}
+
+/// A DP strategy using the discrete gaussian distribution.
+pub struct DiscreteGaussianDpStrategy<B>
+where
+    B: DifferentialPrivacyBudget,
+{
+    budget: B,
+}
+
+/// A DP strategy using the discrete gaussian distribution providing zero-concentrated DP.
+pub type ZCdpDiscreteGaussian = DiscreteGaussianDpStrategy<ZCdpBudget>;
+
+impl DifferentialPrivacyStrategy for DiscreteGaussianDpStrategy<ZCdpBudget> {
+    type Budget = ZCdpBudget;
+    type Distribution = DiscreteGaussian;
+    type Sensitivity = Ratio<BigUint>;
+
+    fn from_budget(budget: ZCdpBudget) -> DiscreteGaussianDpStrategy<ZCdpBudget> {
+        DiscreteGaussianDpStrategy { budget }
+    }
+
+    /// Create a new sampler from the Discrete Gaussian Distribution with a standard
+    /// deviation calibrated to provide `1/2 epsilon^2` zero-concentrated differential
+    /// privacy when added to the result of an integer-valued function with sensitivity
+    /// `sensitivity`, following Theorem 4 from [[CKS20]]
+    ///
+    /// [CKS20]: https://arxiv.org/pdf/2004.00010.pdf
+    fn create_distribution(
+        &self,
+        sensitivity: Ratio<BigUint>,
```

Comment:
I realized that it is common to have irrational sensitivities (e.g. sqrt(2) for the L2 sensitivity of a histogram under Substitution-DP), which is problematic here. 

I think that the code only uses `sigma.pow(2)` and never `sigma` itself, so it might be worth taking `sensitivity_square` or `variance` as an argument at some point. We can always use rational approximations in the meantime.

Reply:
we need the floor of the square root of `sigma` for the discrete gaussian sampler. [it can be computed exactly](https://github.com/IBM/discrete-gaussian-differential-privacy/blob/cb190d2a990a78eff6e21159203bc888e095f01b/discretegauss.py#L101) , do you think i should do that?

Reply:
Great point! I think it's worth implementing that, so we don't have to worry about irrational numbers later on, but I also don't want to delay this PR. Up to you, @ooovi and @divergentdave.

Reply:
Addressing this in a follow-up makes sense to me. Rounding the sensitivity up to a rational number is a very close workaround in the meantime.

Reply:
I suggest filing an issue to track work.